File: leetcode/20.py
"""
Runtime: 28 ms, faster than 84.17% of Python3 online submissions for Valid Parentheses.
Memory Usage: 14.1 MB, less than 87.54% of Python3 online submissions for Valid Parentheses.
"""


class Solution:
    def isValid(self, s: str) -> bool:
        # pop results off the stack
        cl_to_op = {")": "(", "]": "[", "}": "{"}
        s, res = list(s), ""
        for c in s:
            if c in cl_to_op.keys():
                if not res or res and not cl_to_op[c] == res[-1]:
                    return False
                res = res[:-1]
            elif c in cl_to_op.values():
                res += c

        return not res

File: leetcode/20.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Valid Parentheses.
// Memory Usage: 6.2 MB, less than 96.09% of C++ online submissions for Valid Parentheses.
class Solution {
 public:
  bool isValid(string s) {
    // push char onto the stack
    // while closing pair, remove that pair
    vector<char> stack = {};

    auto match = [=](char c) {
      if (c == ')') {
        return '(';
      } else if (c == '}') {
        return '{';
      } else if (c == ']') {
        return '[';
      } else {
        return '@';
      }
    };
    for (char c : s) {
      char curr = c;
      if (!stack.empty() && match(curr) == stack[stack.size() - 1]) {
        stack.pop_back();
        curr = !stack.empty() ? stack[stack.size() - 1] : '@';
      } else {
        stack.push_back(c);
      }
    }

    return stack.size() == 0;
  }
};
